viewing_party/party.py

```diff
@@ -1,23 +1,276 @@
+    
+    if len(user_data["watched"]) < 1:
+        watched_avg_rating = 0.0 
+        return watched_avg_rating
+
+    for movie in user_data["watched"]:
+        total_ratings += movie["rating"]
+
+    watched_avg_rating = total_ratings / num_of_movies
+
+    return watched_avg_rating
+
+def get_most_watched_genre(user_data):
+    fantasy_count = 0
+    action_count = 0
+    intrigue_count = 0
+    
+    if len(user_data["watched"]) < 1:
+        most_watched_genre = None 
+        return most_watched_genre
+
+    for movie in user_data["watched"]:
+        if movie["genre"] == "Fantasy":
+            fantasy_count += 1
+        if movie["genre"] == "Action":
+            action_count += 1
+        if movie["genre"] == "Intrigue":
+            intrigue_count += 1
+
+    if fantasy_count > action_count and fantasy_count > intrigue_count:
+        most_watched_genre = "Fantasy"
+    if action_count > fantasy_count and action_count > intrigue_count:
+        most_watched_genre = "Action"
+    if intrigue_count > action_count and intrigue_count > fantasy_count:
+        most_watched_genre = "Intrigue"
+    return most_watched_genre
+
+
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+
+def get_unique_watched(user_data):
+    user_watched_list = []
+    friend_watched_list = []
+
+    for movie in user_data["watched"]:
+        user_watched_list.append(movie["title"])
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_watched_list.append(movie["title"])
+    
+    user_watched_set = set(user_watched_list)
+    friend_watched_set = set(friend_watched_list)
+    user_unique_watched_set = user_watched_set.difference(friend_watched_set)
+    user_unique_watched_list = []
+
+    for unique_movie in user_unique_watched_set:
+        for movie in user_data["watched"]:
+            if movie["title"] == unique_movie:
+                user_unique_watched_list.append(movie)
+
+    return user_unique_watched_list
+
+def get_friends_unique_watched(user_data):
+    user_watched_list = []
+    friend_watched_list = []
+
+    for movie in user_data["watched"]:
+        user_watched_list.append(movie["title"])
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_watched_list.append(movie["title"])
+    
+    user_watched_set = set(user_watched_list)
+    friend_watched_set = set(friend_watched_list)
+    friend_unique_watched_set = friend_watched_set.difference(user_watched_set)
+    friend_unique_watched_list = []
+
+    for unique_movie in friend_unique_watched_set:
+        for friend in user_data["friends"]:
+            for movie in friend["watched"]:
+                if movie not in friend_unique_watched_list:
+                    if movie["title"] == unique_movie:
+                        friend_unique_watched_list.append(movie)

```

Comment:
The general approach looks fine, but you're right, we could turn this around a bit, similar to `get_unique_watched` above.

```python
    for friend in user_data["friends"]:
        for movie in friend["watched"]:
            if movie["title"] in friend_unique_watched_set:
                friend_unique_watched_list.append(movie)
                friend_unique_watched_set.remove(movie["title"])
```